jsonrpc/filter_manager.go

```diff
@@ -342,6 +349,116 @@ func (f *FilterManager) Exists(id string) bool {
 	return ok
 }
 
+// GetLogsForQuery return array of logs for given query
+func (f *FilterManager) GetLogsForQuery(query *LogQuery) ([]*Log, error) {
+	result := make([]*Log, 0)
+	parseReceipts := func(block *types.Block) error {
+		receipts, err := f.store.GetReceiptsByHash(block.Header.Hash)
+		if err != nil {
+			return err
+		}
+
+		for indx, receipt := range receipts {
+			for logIndx, log := range receipt.Logs {
+				if query.Match(log) {
+					result = append(result, &Log{
+						Address:     log.Address,
+						Topics:      log.Topics,
+						Data:        argBytes(log.Data),
+						BlockNumber: argUint64(block.Header.Number),
+						BlockHash:   block.Header.Hash,
+						TxHash:      block.Transactions[indx].Hash,
+						TxIndex:     argUint64(indx),
+						LogIndex:    argUint64(logIndx),
+					})
+				}
+			}
+		}
+
+		return nil
+	}
+
+	if query.BlockHash != nil {
+		block, ok := f.store.GetBlockByHash(*query.BlockHash, true)
+		if !ok {
+			return nil, fmt.Errorf("not found")
+		}
+
+		if len(block.Transactions) == 0 {
+			// no txs in block, return empty response
+			return result, nil
+		}
+
+		if err := parseReceipts(block); err != nil {
+			return nil, err
+		}
+
+		return result, nil
+	}
+
+	head := f.store.Header().Number
+
+	resolveNum := func(num BlockNumber) uint64 {
+		if num == PendingBlockNumber {
+			num = LatestBlockNumber
+		}
+
+		if num == EarliestBlockNumber {
+			num = 0
+		}
+
+		if num == LatestBlockNumber {
+			return head
+		}
+
+		return uint64(num)
+	}
+
+	from := resolveNum(query.fromBlock)
+	to := resolveNum(query.toBlock)
+
+	if to < from {
+		return nil, fmt.Errorf("incorrect range")
+	}
+
+	for i := from; i <= to; i++ {
+		block, ok := f.store.GetBlockByNumber(i, true)
+		if !ok {
+			break
+		}
+
+		if block.Header.Number == 0 || len(block.Transactions) == 0 {
+			// do not check logs in genesis and skip if no txs
+			continue
+		}
```

Comment:
Why even fetch the block if you're gonna just continue if the block is the genesis block?

You know the number of the block before you even fetch it (i), so you should perform this check sooner (if it's genesis, continue)

Reply:
Fixed fa99b5d16e185f74fc9c216e6578d223e8145ad1
Fixed Linter 9d46530204c5507e2451aa0d2b09c0cc34713b36